src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -252,6 +260,43 @@ protected HttpMessage prepareRequestMessage(UsernamePasswordAuthenticationCreden
 			return requestMessage;
 		}
 
+		private void replaceAntiCsrfTokenAndCookieValueIfRequired(HttpMessage requestMessage) throws URIException, NullPointerException, HttpMalformedHeaderException {
+			if(extAntiCsrf == null) {
+				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
+			}
+			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			if(extAntiCsrf != null) {
+				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
+			}
+			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
+			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
+				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+			}
+			
+			if (antiCsrfTokenOfFirstLoginMsg != null) {
+				this.requireAnitCsrfTokenWhenLogin = true;
+				URI loginURI = new URI(loginRequestURL, true, "UTF-8");
```

Comment:
Shouldn't this use the URL from `antiCsrfTokenOfFirstLoginMsg`? (It might not be the same.)

Reply:
`antiCsrfTokenOfFirstLoginMsg` is got from passing the `HttpMessage` of `loginSiteNode` to the `extAntiCsrf.getToken(HttpMessage)`.
as URL of `antiCsrfTokenOfFirstLoginMsg` is referring to the `POST` method and I want a fresh 'GET`, I opted to this. But as this is just URI, I can change this. Did not get the point where URL created using `loginRequestURL` and URL from `antiCsrfTokenOfFirstLoginMsg` may not be the same. Any example?

Reply:
The `antiCsrfTokenOfFirstLoginMsg` should point to a GET not POST, could you confirm that?

That happens when the login form is not submitted to the same page.

Reply:
sorry for the misunderstanding, as `loginSiteNode` is referring to `POST` and I got `antiCsrfTokenOfFirstLoginMsg` from the history reference of `loginSiteNode` I thought the `antiCsrfTokenOfFirstLoginMsg` also pointing to `POST`.

`antiCsrfTokenOfFirstLoginMsg` is pointing to `GET`.
![image](https://user-images.githubusercontent.com/17596394/40320700-2c905dd6-5d4a-11e8-83f9-b6dbecf76dbb.png)